### ISSUE TITLE
JPAConverter may cause DataTypeException if it does not detect the proper <T, U> types using reflection

### ISSUE THREAD

**Issue by kevinzogg** [jOOQ/jOOQ#10951]
### Your question:

We are trying to upgrade from 3.12.4 to 3.14.3 and there is an issue I'm not sure if it's intended, a configuration problem or even a bug.

We have a field with an AttributeConverter:

```
@Column(name = "product_name_i18n", nullable = false)
@Convert(converter = TranslationMapAttributeConverter.class)
private TranslationMap productNameI18n;
```

The generated code looks fine, including the converter:

```
/**
 * The column <code>BIL_PRODUCT.PRODUCT_NAME_I18N</code>.
 */
public final TableField<BilProductRecord, TranslationMap> PRODUCT_NAME_I18N = createField(DSL.name("PRODUCT_NAME_I18N"), org.jooq.impl.SQLDataType.VARCHAR(255).nullable(false), this, "", new org.jooq.impl.JPAConverter(ch.jobtool.model.common.TranslationMapAttributeConverter.class));
```
Also, the generated Record has the correct Type:
```
public class BilProductRecord extends UpdatableRecordImpl<BilProductRecord> implements Record11<Long, Boolean, Boolean, TranslationMap, ...
```
The issue arises when we try to fetchInto(...). jOOQ tries to find a Converter (Object <-> TranslationMap):
```
Caused by: org.jooq.exception.DataTypeException: No Converter found for types java.lang.Object and xyz.TranslationMap
```

I dug through the code and found that in the DefaultRecordMapper there is a private map method (603) where primitive types are handled, and otherwise it calls `record.get(index, mType);`. Where mType is TranslationMap. This then leads to the following Code:

```
public final <T> T get(int index, Class<? extends T> type) {
    return Tools.converterOrFail(this, this.field(this.safeIndex(index)).getType(), type).from(this.get(index));
}
```
This is when I found out that the Type of the field seems to be Object, whereas I would have expected TranslationMap. Here jOOQ tries to find a converter from Object to TranslationMap. I am very much confused, since the value should already be of type TranslationMap. What do I miss here?

For the record: `PRODUCT_NAME_I18N.getType()` does indeed return Object. Is this intended or is it something we configured wrong for the generated code?

Note: this all worked fine with jOOQ 3.12.4.

### Versions:

- jOOQ:  3.14.3
- Java: 11.0.2

**Comment by lukaseder** [comment]
Thanks for your message.

Can you show a complete example of how you build a query and call `fetchInto()`, as well as the code for your various types, including `TranslationMap` and `TranslationMapAttributeConverter`?

This may well be a bug, but there are many things that play into this, so perhaps an MCVE might be helpful: https://github.com/jOOQ/jOOQ-mcve

**Comment by kevinzogg** [comment]
I was able to create a MCVE: https://github.com/kevinzogg/jOOQ-mcve/tree/feature/issue-10951

I had to set up the project different than in your linked repository, so there is not much in common anymore. But I don't think that's a problem.

Maybe some key Elements are:

* A module with the JPA Entity and the aforementioned TranslationMap.
* Code Generation from this entity (https://www.jooq.org/doc/latest/manual/code-generation/codegen-jpa/)
* Testdatabase setup with liquibase (as in our project)
* The failing test to see the issue is in the persistence module: ProductRepositoryTest

**Comment by lukaseder** [comment]
Thanks a lot for taking the time to do that. I'll look into your MCVE right away

**Comment by lukaseder** [comment]
I can reproduce the issue with the MCVE from the command line using `mvn install`, but not from within Eclipse. Let's see if this is some classpath issue...

**Comment by lukaseder** [comment]
My bad. It reproduces in Eclipse as well...

**Comment by lukaseder** [comment]
... but not always :)

**Comment by lukaseder** [comment]
Well, the workaround is to implement a `ConverterProvider` and register that with your `Configuration` for this type configuration.

**Comment by kevinzogg** [comment]
Yes, along the lines of what is mentioned in https://www.jooq.org/doc/3.15/manual/sql-execution/fetching/converter-provider/

That is possible (and works). Handling the Object -> TranslationMap conversion just seems strange, since it takes Object and basically just tries to Cast it to TranslationMap.

**Comment by lukaseder** [comment]
The regression was introduced by #10071 in 3.14.0

**Comment by lukaseder** [comment]
This method in `AbstractRecord` is wrong:

```java
    @Override
    public final <T> T get(int index, Class<? extends T> type) {
        return (T) converterOrFail(this, field(safeIndex(index)).getType(), (Class) type).from(get(index));
    }
```

There's no need for a converter if `type.isAssignableFrom(get(index))`

**Comment by lukaseder** [comment]
It's curious, though, that `field(safeIndex(index)).getType()` resolves to `Object`, and not to `TranslationMap`. Something's going wrong with the reflection in `JPAConverter`

**Comment by lukaseder** [comment]
I.e. the types cannot be resolved correctly:

![image](https://user-images.githubusercontent.com/734593/99403668-028aab80-28eb-11eb-9cfd-2433e2b3026b.png)

**Comment by lukaseder** [comment]
Ah, I get it. We're finding the actual methods first, not the bridge methods, which we're looking for...

![image](https://user-images.githubusercontent.com/734593/99403901-48e00a80-28eb-11eb-8dc5-d1d9aef502b4.png)

**Comment by lukaseder** [comment]
Yep, that fixes it. It also explains the flakiness. Relying on the order of returned values in the reflection API is never a good idea. The `JPAConverter` worked "by accident" given that it may or may not have produce the correct `<T>` and `<U>` types depending on how the classes were loaded / instrumented / etc.

Fixed in jOOQ 3.15.0 and 3.14.4 (#10960). Thanks again for your thorough report!

**Comment by lukaseder** [comment]
The solution probably doesn't work in case the `AttributeConverter` implementation is part of some type hierarchy, in case of which the actual `Class` may not contain the declared method we're interested in. If we continued to use `Class::getMethods`, then we might still produce flaky results in case there are several bridge methods in the hierarchy, each with some more type refinement.

I'm happy to keep this as a "known limitation" rather than overengineering this feature...
